File: JS/BOJ-S3-1213-Make_Palindrome.js
'use strict';

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
let input = fs.readFileSync(filePath).toString().trim();

let output = [];
const solution = (input) => {
  const cnt = Array.from({ length: 26 }, () => 0);
  for (let i = 0; i < input.length; ++i) {
    cnt[input.charCodeAt(i) - 65]++;
  }
  let mid = '';
  let flag = 0;
  let tmp = '';
  for (let i = cnt.length - 1; i >= 0; --i) {
    if (cnt[i] & 1) {
      mid = String.fromCharCode(i + 65);
      --cnt[i];
      ++flag;
    }
    if (flag > 1) break;
    for (let j = 0; j < cnt[i] / 2; ++j) {
      let s = String.fromCharCode(i + 65);
      tmp = s + tmp;
      tmp += s;
    }
  }
  if (mid)
    tmp =
      tmp.substring(0, tmp.length / 2) + mid + tmp.substring(tmp.length / 2);

  flag > 1 ? output.push("I'm Sorry Hansoo") : output.push(tmp);
};

solution(input);
fs.writeFileSync('/dev/stdout', output.join('').trimEnd());
// fs.writeFileSync(
//   '../output.txt',
//   output.join('').trimEnd(),
//   (err) => console.error
// );

File: CPP/BOJ-S3-1213-Make_Palindrome.cpp
#include <bits/stdc++.h>
using namespace std;

string s, ret;
int cnt[26], flag;
char mid;

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  // freopen("input.txt", "r", stdin);
  // freopen("output.txt", "w", stdout);

  cin >> s;
  for (char a : s)
  {
    cnt[a - 65]++;
  }

  for (int i = 25; i >= 0; --i)
  {
    if (cnt[i] & 1)
    {
      mid += char(i + 65);
      ++flag;
      --cnt[i];
    }
    if (flag == 2)
    {
      cout << "I'm Sorry Hansoo" << '\n';
      exit(0);
    }
    for (int j = 0; j < cnt[i] / 2; ++j)
    {
      ret = char(i + 65) + ret;
      ret += char(i + 65);
    }
  }

  if (mid)
    ret.insert(ret.begin() + ret.size() / 2, mid);

  cout << ret << '\n';

  return 0;
}